test/common_distributed.py

```diff
@@ -0,0 +1,165 @@
+        return wrapper
+
+    return decorator
+
+
+def skip_for_known_issues(func):
+    """Skips a test due to known issues (for c10d)."""
+    @wraps(func)
+    def wrapper(*args, **kwargs):
+        sys.exit(TEST_SKIPS['known_issues'].exit_code)
+
+    return wrapper
+
+
+def requires_gloo():
+    return unittest.skipUnless(
+        c10d.is_gloo_available(),
+        "c10d was not compiled with the Gloo backend",
+    )
+
+
+def requires_nccl():
+    return unittest.skipUnless(
+        c10d.is_nccl_available(),
+        "c10d was not compiled with the NCCL backend",
+    )
+
+
+def requires_mpi():
+    return unittest.skipUnless(
+        c10d.is_mpi_available(),
+        "c10d was not compiled with the MPI backend",
+    )
+
+
+TIMEOUT_DEFAULT = 30
+TIMEOUT_OVERRIDE = {}
+
+
+def get_timeout(test_id):
+    return TIMEOUT_OVERRIDE.get(test_id.split('.')[-1], TIMEOUT_DEFAULT)
+
+
+class MultiProcessTestCase(TestCase):
+    MAIN_PROCESS_RANK = -1
+
+    @property
+    def world_size(self):
+        return 4
+
+    @staticmethod
+    def join_or_run(fn):
+        @wraps(fn)
+        def wrapper(self):
+            if self.rank == self.MAIN_PROCESS_RANK:
+                self._join_processes(fn)
+            else:
+                fn(self)
+        return wrapper
+
+    # The main process spawns N subprocesses that run the test.
+    # This function patches overwrites every test function to either
+    # assume the role of the main process and join its subprocesses,
+    # or run the underlying test function.
+    @classmethod
+    def setUpClass(cls):
+        for attr in dir(cls):
+            if attr.startswith('test'):
+                fn = getattr(cls, attr)
+                setattr(cls, attr, cls.join_or_run(fn))
+
+    def setUp(self):
+        super(MultiProcessTestCase, self).setUp()
+        self.rank = self.MAIN_PROCESS_RANK
+        self.file = tempfile.NamedTemporaryFile(delete=False)
+        self.processes = [self._spawn_process(rank) for rank in range(int(self.world_size))]
+
+    def tearDown(self):
+        super(MultiProcessTestCase, self).tearDown()
+        for p in self.processes:
+            p.terminate()
+
+    def _spawn_process(self, rank):
+        name = 'process ' + str(rank)
+        process = multiprocessing.Process(target=self._run, name=name, args=(rank,))
+        process.start()
+        return process
+
+    def _run(self, rank):
+        self.rank = rank
+
+        # self.id() == e.g. '__main__.TestDistributed.test_get_rank'
+        # We're retreiving a corresponding test and executing it.
+        getattr(self, self.id().split(".")[2])()
+        sys.exit(0)
+
+    def _join_processes(self, fn):
+        timeout = get_timeout(self.id())
+        start_time = time.time()
+        for p in self.processes:
+            p.join(timeout)
+        elapsed_time = time.time() - start_time
+        self._check_return_codes(elapsed_time)
+
+    def _check_return_codes(self, elapsed_time):
```

Comment:
Have you considered using, `torch.multiprocessing.SpawnContext.join()`?

`SpawnContext` is like a "Process Group". It monitors the health of all the child processes.
It reports any Python exception or crash happend in child processes. (Proper error propogation).

It also handles "out of order termination".
Say if self.processes[2] crashes, self.processes[0] hangs,
this code
```
        for p in self.processes:
            p.join()
```
would hang.

While, `torch.multiprocessing.SpawnContext.join()` resolves this, by listening (multiprocessing.connection.wait(sentinel_list, timeout=None)) to sentinel objects of all processes and response to the first process that has a ready(could be succeded or failed) status, and kills the rest children is necessary.

Reply:
@xsh6528 I thought if we use `spawn`, we cannot pickle `TestCase` objects (basically `self`) and pass it to subprocesses? It leads to the test implementation in #23228, where we have to use either `@staticmethod` or `@classmethod` and have two functions for each test.  Or is there anyway to get around that?

Reply:
@mrshenli 

The problem is that **spawned** children don't have the file descriptors created in their parent. 
So Python has to pickle the FileIO object (`self.file`), in order to pass it to the copies of `self` in children.

However, **forked** children have all the fds their parent have, so there is no need to pickle in this case.

I tried to change `mp = multiprocessing.get_context('spawn')` to `mp = multiprocessing.get_context('fork')`. It works!
But it's likely we don't want to change `torch.multiprocessing.spawn` at this point.

Another way is to 
```
import uuid

class MultiProcessTestCase(TestCase):
    def setUp(self):
        # self.file = tempfile.NamedTemporaryFile(delete=False)  # instead of this way
        self.init_method = "file://tmp/{random_name}".format(uund.uuid4())

def _init_rpc(func):
    def wrapper(self):
        store = dist.FileStore(self.file.name, self.world_size)
        dist.init_process_group(backend='gloo', rank=self.rank,
                                world_size=self.world_size, init_method=self.init_method)
        dist.init_rpc('worker%d' % self.rank)
        func(self)

    return wrapper
```

But I don't know if you think `init_method=` is an API to be deprecated.

Reply:
@xush6528 sorry I am little lost, what are the suggested changes here?

Reply:
@mrshenli 

Sorry for the confusion. 

The line I was referring to is 

https://github.com/pytorch/pytorch/blob/e982e46de34eed411a11c0b7ca034a5b7b177fa4/torch/multiprocessing/spawn.py#L152

The observation is that using "fork" instead of "spawn" can avoid "pickling".

But I don't think this line should be changed, because we don't want "fork" to bring in copies of resources in the parent process .

So actually I don't have a suggested change. I think this `MultiProcessTestCase` implementation, modifying all test methods in `setUpClass`, is fairely reasonble.

Reply:
Got it, thanks!